klippy/extras/bltouch.py

```diff
@@ -53,54 +59,96 @@ def __init__(self, config):
-            raise homing.EndstopError("BLTouch sensor test failed")
+            raise homing.EndstopError("BLTouch failed to %s" % (msg,))
         for s, pos in zip(self.mcu_endstop.get_steppers(), prev_positions):
             s.set_commanded_position(pos)
+    def raise_probe(self):
+        self.send_cmd('reset')
+        check_start_time = self.send_cmd('pin_up', duration=self.pin_move_time)
+        check_end_time = self.send_cmd(None)
+        self.verify_state(check_start_time, check_end_time,
+                          self.pin_up_triggered, "raise probe")
+    def test_sensor(self):
+        if not self.pin_up_touch_triggered or self.pin_up_triggered:
+            # Nothing to test
+            return
+        toolhead = self.printer.lookup_object('toolhead')
+        print_time = toolhead.get_last_move_time()
+        if print_time < self.next_test_time:
+            self.next_test_time = print_time + TEST_TIME
+            return
+        # Raise the bltouch probe and test if probe is raised
+        self.sync_print_time()
+        check_start_time = self.send_cmd('reset', duration=self.pin_move_time)
+        check_end_time = self.send_cmd('touch_mode')
+        self.send_cmd(None)
+        self.verify_state(check_start_time, check_end_time, True,
+                          "verify sensor state")
         # Test was successful
-        self.next_test_time = home_time + TEST_TIME
-        toolhead.reset_print_time(home_time + 2. * MIN_CMD_TIME)
+        self.next_test_time = check_end_time + TEST_TIME
+        self.sync_print_time()
     def home_prepare(self):
         self.test_sensor()
-        toolhead = self.printer.lookup_object('toolhead')
-        print_time = toolhead.get_last_move_time()
-        self.send_cmd(print_time, 'pin_down')
-        self.send_cmd(print_time + self.pin_move_time, 'touch_mode')
-        toolhead.dwell(self.pin_move_time + MIN_CMD_TIME)
+        self.sync_print_time()
+        self.send_cmd('pin_down', duration=self.pin_move_time - MIN_CMD_TIME)
+        self.send_cmd(None)
+        self.sync_print_time()
         self.mcu_endstop.home_prepare()
+        self.start_mcu_pos = [(s, s.get_mcu_position())
+                              for s in self.mcu_endstop.get_steppers()]
     def home_finalize(self):
-        toolhead = self.printer.lookup_object('toolhead')
-        print_time = toolhead.get_last_move_time()
-        self.send_cmd(print_time, 'reset')
-        self.send_cmd(print_time + MIN_CMD_TIME, 'pin_up')
-        self.send_cmd(print_time + MIN_CMD_TIME + self.pin_move_time, None)
-        toolhead.dwell(self.pin_move_time + MIN_CMD_TIME)
+        self.sync_mcu_print_time()
+        self.raise_probe()
+        self.sync_print_time()
+        # Verify the probe actually deployed during the attempt
+        for s, mcu_pos in self.start_mcu_pos:
+            if s.get_mcu_position() == mcu_pos:
+                raise homing.EndstopError("BLTouch failed to deploy")
```

Comment:
I don't understand what you are reporting here.

The user will not receive an "Internal error on command" message.

The code does propagate an error message which will cancel any scripts and inform OctoPrint (or similar) of the error which will cancel any g-code print.  If the user chooses to ignore the error, then that is their choice.

Reply:
I'm reporting that I would like the printer to shut down if there are fatal errors with the probe here. We can't guarantee the state of the pin, so if there are any kinds of movement, it can break the very fragile probe.

I certainly don't want to rely on Octoprint being properly configured to do it for me. 

Fatal error in this case means that even after a reset, and a commanded pin move up the probe still returns an error.

And my following up question was. How do you make this error to be shown, without raising a gcode error, since we don't know if we are running gcode here or not? And without having to modify all gcode commands to handle some custom expection and without modifying `gcode.py`

Reply:
Realistically, what are you going to do if the probe still reports an error after the following two commands?

1. `reset`
2. `pin_up`

Most users would want to run `M112` and report the bug. Some users will have forgot to run `M112` and the relevant information is now gone. 

It is probably safe to not shut down the printer since, the steppers should get disabled. But if you are running a print job and it's not automatically cancelled, then your log will most likely be filled and the actual error gone, and more than 100 commands run, so `M112` will not help you anymore.

Some users notice a loose cable, and have to shut down the printer to fix it anyway.

Some users don't care about the error and tries again, doing another `G28`. However some might have a too low retract distance to take a lowered pin into account, and now you have some angry users with a broken probe.

The only users that are served by keeping the printer up running, is the ones that have some filament stuck on the probe, which prevents it from moving up. But in that case a heated nozzle have probably crashed into the bed, so leaving the printer on with heating enabled looks like a pretty dangerous thing to do...

The absolutely best thing to do, which is what I was planning for my next iteration, would be to retract the z-axis, and then shut the printer off, to prevent the nozzle from being in contact with the bed for too long.

**Regarding the exception handling:**

So in my pull request, the correct thing to do would have been to raise an `EndstopError`, instead of `mcu.Error`? And in the `connect` handler translate that to `mcu.Error` or `config_error`, like you do, which does not make sense to me, because we are talking about a hardware error, which may or may not be caused by misconfiguration. 

I still think that handling all the possible exceptions in the actual command handler is very fragile. For example `cmd_BED_MESH_CALIBRATE` does not handle `EndstopError` currently, at least not directly, so we most likely have bugs there.  And the same for bed tilt calibration. And even if we fix those now, we still still need to document that anything that can possible touch the endstop has to catch this exception.

I also don't agree with that that we know that `home_finalize` is called from a gcode handler. That might be the case right now, but it might not be in the future. And code that assumes something about external more high level code around it is generally very fragile. Imagine for example some kind of filament runout sensor that automatically changes the filament, perhaps does some kind of toolchange, and need an automatic homing at a specific safe location afterwards. That kind of thing could do homing operations outside of gcode. I know it's a quite far fetched case, but my point is that with code you can never assume that something will hold now and forever, unless it's clearly documented, and even then it should be continuously tested, preferably statically.

Finally, I don't agree that the homing code should know how to shutdown the printer itself, that clearly breaks the single responsibility principle. That's why I used two categories of exceptions, `homing.EndstopError` for recoverable errors, and `mcu.Error` for fatal unrecoverable ones. But the problem in that reasoning was that the rest of the code did not know how to report those errors, although it handled them the way it should. So I just made a handler that printed it in a nicer form.

Furthermore my code is not the only code that leaks `mcu.Error`, for example this https://github.com/KevinOConnor/klipper/issues/1027, would have benefited from the change that I made, there would have been a more readable error message. Note that I didn't change any behaviour, I just made mcu.Error print more nicely to the console.

But despite all those opinions of mine, I feel that I wasn't given a clear answer how you exactly liked it to be, so that's why I couldn't make any changes.